core/src/main/java/org/elasticsearch/index/shard/IndexingOperationListener.java

```diff
@@ -38,12 +38,17 @@
     }
 
     /**
-     * Called after the indexing operation occurred.
+     * Called after the indexing operation occurred. Implementations should
+     * check {@link Engine.IndexResult#hasFailure()} for operation failures
```

Comment:
I don't think we need to say this? how about "note that this method is also called when indexing a document didn't succeed because of document related failures. See {@link #postIndex(..)} for engine level failures.